Bug in Quarkus bytecode generation

### Describe the bug

Hi all, I'm Ronald, the creator of [JobRunr](https://github.com/jobrunr/jobrunr), a Quartz replacement.

JobRunr also integrates with Quarkus and to do so, I created a custom Quarkus Extension. I'm currently struggling with this extension.

In this extension, I've written a [RecurringJobsFinder](https://github.com/jobrunr/jobrunr/blob/bugfix/689/framework-support/jobrunr-quarkus-extension/deployment/src/main/java/org/jobrunr/quarkus/extension/deployment/RecurringJobsFinder.java) that searches for all methods annotated with @Recurring and schedules them using a recorder so that the bytecode is recorded.

This is all tested by means of integration tests here:
https://github.com/jobrunr/jobrunr/blob/bugfix/689/framework-support/jobrunr-quarkus-extension/tests/src/main/java/org/jobrunr/quarkus/it/TestService.java#L20-L25

All of this mostly works fine but for one of these recurring jobs, I receive the following exception:

```java
java.lang.RuntimeException: Failed to start quarkus
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
    at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:625)
    at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:696)
    at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
       ...
Caused by: java.lang.RuntimeException: Failed to start quarkus
    at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
    at io.quarkus.runtime.Application.start(Application.java:101)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:253)
    at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:250)
    at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:592)
    at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:640)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$12(ClassBasedTestDescriptor.java:395)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:395)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:211)
    at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:84)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
    ... 52 more
Caused by: java.lang.UnsupportedOperationException
    at java.base/java.util.Collections$UnmodifiableCollection.add(Collections.java:1060)
    at io.quarkus.deployment.steps.JobRunrExtensionProcessor$findRecurringJobAnnotationsAndScheduleThem1351984528.deploy_0(Unknown Source)
    at io.quarkus.deployment.steps.JobRunrExtensionProcessor$findRecurringJobAnnotationsAndScheduleThem1351984528.deploy(Unknown Source)
    ... 68 more
```

The bytecode generated by the `Recorder` is as follows (thanks @gastaldi!):
```java
    public void deploy_0(StartupContext var1, Object[] var2) {
        JobRunrRecurringJobRecorder var4 = new JobRunrRecurringJobRecorder();
        ArrayList var3 = new ArrayList();
        JobDetails var6 = new JobDetails("org.jobrunr.quarkus.it.TestService", (String)null, "aRecurringJob", (List)var3);
        Object var5 = var1.getValue("proxykey93");
        var4.schedule((BeanContainer)var5, "my-recurring-job", var6, "*/15 * * * *", (String)null, (String)null);
        ArrayList var7 = new ArrayList(1);
        JobParameter var8 = new JobParameter("org.jobrunr.jobs.context.JobContext", "org.jobrunr.jobs.context.JobContext", (Object)null);
        ((Collection)var7).add(var8);
        JobDetails var10 = new JobDetails("org.jobrunr.quarkus.it.TestService", (String)null, "anotherRecurringJob", (List)var7);
        ((Collection)var10.getJobParameters()).add(var8);
        Object var9 = var1.getValue("proxykey93");
        var4.schedule((BeanContainer)var9, "another-recurring-job-with-jobContext", var10, (String)null, "PT10M", (String)null);
    }
```

We noticed that `var8` is added twice for some reason. As `((Collection)var10.getJobParameters())` returns an unmodifiable list, the second add fails. But, I do not understand the reason why the parameter is added twice in the first place as that will also cause problems later on in JobRunr (2 parameters where the method only expects 1).

### Expected behavior

The parameter should only be added once.

### Actual behavior

The parameter is added twice.

### How to Reproduce?

1. git clone https://github.com/jobrunr/jobrunr
2. git checkout `bugfix/689`
3. import project in Intellij
4. Run test `./framework-support/jobrunr-quarkus-extension/tests/src/test/java/org/jobrunr/quarkus/it/JobRunrFunctionalityTest.java`


### Output of `uname -a` or `ver`

Darwin Ronalds-MacBook-Pro.local 21.6.0 Darwin Kernel Version 21.6.0: Thu Sep 29 20:13:56 PDT 2022; root:xnu-8020.240.7~1/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "17.0.1" 2021-10-19 OpenJDK Runtime Environment (build 17.0.1+12-39) OpenJDK 64-Bit Server VM (build 17.0.1+12-39, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.15.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.5.1

### Additional information

See also the analysis of @gastaldi at https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/JobRunr.20Quarkus.20extension.20-.20UnsupportedOperationException

--- a/core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java
@@ -1176,8 +1176,10 @@ public class BytecodeRecorderImpl implements RecorderContext {
             if (nonDefaultConstructorHolder.constructor.getParameterCount() > 0) {
                 Parameter[] parameters = nonDefaultConstructorHolder.constructor.getParameters();
                 for (int i = 0; i < parameters.length; ++i) {
-                    String name = parameters[i].getName();
-                    constructorParamNameMap.put(name, i);
+                    if (parameters[i].isNamePresent()) {
+                        String name = parameters[i].getName();
+                        constructorParamNameMap.put(name, i);
+                    }
                 }
             }
         } else if (classesToUseRecordableConstructor.contains(param.getClass())) {
@@ -1199,8 +1201,10 @@ public class BytecodeRecorderImpl implements RecorderContext {
             if (current.getParameterCount() > 0) {
                 Parameter[] parameters = current.getParameters();
                 for (int i = 0; i < parameters.length; ++i) {
-                    String name = parameters[i].getName();
-                    constructorParamNameMap.put(name, i);
+                    if (parameters[i].isNamePresent()) {
+                        String name = parameters[i].getName();
+                        constructorParamNameMap.put(name, i);
+                    }
                 }
             }
         } else {
@@ -1223,8 +1227,10 @@ public class BytecodeRecorderImpl implements RecorderContext {
                 if (selectedCtor.getParameterCount() > 0) {
                     Parameter[] ctorParameters = selectedCtor.getParameters();
                     for (int i = 0; i < ctorParameters.length; ++i) {
-                        String name = ctorParameters[i].getName();
-                        constructorParamNameMap.put(name, i);
+                        if (ctorParameters[i].isNamePresent()) {
+                            String name = ctorParameters[i].getName();
+                            constructorParamNameMap.put(name, i);
+                        }
                     }
                 }
             }

--- a/core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java
@@ -1173,6 +1173,13 @@ public class BytecodeRecorderImpl implements RecorderContext {
                 nonDefaultConstructorHandles[i] = loadObjectInstance(obj, existing,
                         parameterTypes[count++], relaxedValidation);
             }
+            if (nonDefaultConstructorHolder.constructor.getParameterCount() > 0) {
+                Parameter[] parameters = nonDefaultConstructorHolder.constructor.getParameters();
+                for (int i = 0; i < parameters.length; ++i) {
+                    String name = parameters[i].getName();
+                    constructorParamNameMap.put(name, i);
+                }
+            }
         } else if (classesToUseRecordableConstructor.contains(param.getClass())) {
             Constructor<?> current = null;
             int count = 0;
@@ -1191,7 +1198,7 @@ public class BytecodeRecorderImpl implements RecorderContext {
             nonDefaultConstructorHandles = new DeferredParameter[current.getParameterCount()];
             if (current.getParameterCount() > 0) {
                 Parameter[] parameters = current.getParameters();
-                for (int i = 0; i < current.getParameterCount(); ++i) {
+                for (int i = 0; i < parameters.length; ++i) {
                     String name = parameters[i].getName();
                     constructorParamNameMap.put(name, i);
                 }
@@ -1215,7 +1222,7 @@ public class BytecodeRecorderImpl implements RecorderContext {
 
                 if (selectedCtor.getParameterCount() > 0) {
                     Parameter[] ctorParameters = selectedCtor.getParameters();
-                    for (int i = 0; i < selectedCtor.getParameterCount(); ++i) {
+                    for (int i = 0; i < ctorParameters.length; ++i) {
                         String name = ctorParameters[i].getName();
                         constructorParamNameMap.put(name, i);
                     }

--- a/core/deployment/src/test/java/io/quarkus/deployment/recording/BytecodeRecorderTestCase.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/recording/BytecodeRecorderTestCase.java
@@ -215,11 +215,7 @@ public class BytecodeRecorderTestCase {
     @Test
     public void testLargeCollection() throws Exception {
 
-        List<TestJavaBean> beans = new ArrayList<>();
-        for (int i = 0; i < 100000; ++i) {
-            beans.add(new TestJavaBean("A string", 99));
-        }
-
+        List<TestJavaBean> beans = Collections.nCopies(100000, new TestJavaBean("A string", 99));
         runTest(generator -> {
             TestRecorder recorder = generator.getRecordingProxy(TestRecorder.class);
             recorder.list(beans);

--- a/core/deployment/src/test/java/io/quarkus/deployment/recording/BytecodeRecorderTestCase.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/recording/BytecodeRecorderTestCase.java
@@ -1,5 +1,6 @@
 package io.quarkus.deployment.recording;
 
+import static org.assertj.core.api.Assertions.assertThatCode;
 import static org.junit.jupiter.api.Assertions.assertArrayEquals;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
@@ -135,6 +136,30 @@ public class BytecodeRecorderTestCase {
             TestRecorder recorder = generator.getRecordingProxy(TestRecorder.class);
             recorder.bean(new TestJavaBeanSubclass("A string", 99, "PUT"));
         }, new TestJavaBeanSubclass("A string", 99, "PUT"));
+    }
+
+    @Test
+    public void testJobDetails() throws Exception {
+        runTest(generator -> {
+            assertThatCode(() -> {
+                generator.registerNonDefaultConstructor(
+                        JobParameter.class.getDeclaredConstructor(String.class, String.class, Object.class),
+                        jobParameter -> Arrays.asList(
+                                jobParameter.getClassName(),
+                                jobParameter.getActualClassName(),
+                                jobParameter.getObject()));
+                generator.registerNonDefaultConstructor(
+                        JobDetails.class.getDeclaredConstructor(String.class, String.class, String.class, List.class),
+                        jobDetails -> Arrays.asList(
+                                jobDetails.getClassName(),
+                                jobDetails.getStaticFieldName(),
+                                jobDetails.getMethodName(),
+                                jobDetails.getJobParameters()));
+            }).doesNotThrowAnyException();
+            TestRecorder recorder = generator.getRecordingProxy(TestRecorder.class);
+            recorder.bean(new JobDetails("A string", null, "methodName", List.of(JobParameter.JobContext)));
+        }, new JobDetails("A string", null, "methodName", List.of(JobParameter.JobContext)));
+
     }
 
     @Test

--- a/core/deployment/src/test/java/io/quarkus/deployment/recording/JobDetails.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/recording/JobDetails.java
@@ -1,0 +1,88 @@
+package io.quarkus.deployment.recording;
+
+import static java.util.Collections.unmodifiableList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class JobDetails {
+
+    private final String className;
+    private final String staticFieldName;
+    private final String methodName;
+    private final ArrayList<JobParameter> jobParameters;
+    private Boolean cacheable;
+
+    private JobDetails() {
+        this(null, null, null, null);
+        // used for deserialization
+    }
+
+    public JobDetails(String className, String staticFieldName, String methodName, List<JobParameter> jobParameters) {
+        this.className = className;
+        this.staticFieldName = staticFieldName;
+        this.methodName = methodName;
+        this.jobParameters = new ArrayList<>(jobParameters);
+        this.cacheable = false;
+    }
+
+    public String getClassName() {
+        return className;
+    }
+
+    public String getStaticFieldName() {
+        return staticFieldName;
+    }
+
+    public boolean hasStaticFieldName() {
+        return staticFieldName != null;
+    }
+
+    public String getMethodName() {
+        return methodName;
+    }
+
+    public List<JobParameter> getJobParameters() {
+        return unmodifiableList(jobParameters);
+    }
+
+    public Class[] getJobParameterTypes() {
+        return jobParameters.stream()
+                .map(JobParameter::getClassName)
+                .toArray(Class[]::new);
+    }
+
+    public Object[] getJobParameterValues() {
+        return jobParameters.stream()
+                .map(JobParameter::getObject)
+                .toArray();
+    }
+
+    public Boolean getCacheable() {
+        return cacheable;
+    }
+
+    public void setCacheable(boolean cacheable) {
+        this.cacheable = cacheable;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (!(o instanceof JobDetails))
+            return false;
+        JobDetails that = (JobDetails) o;
+        return Objects.equals(className, that.className)
+                && Objects.equals(staticFieldName, that.staticFieldName)
+                && Objects.equals(methodName, that.methodName)
+                && Objects.equals(jobParameters, that.jobParameters)
+                && Objects.equals(cacheable, that.cacheable);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(className, staticFieldName, methodName, jobParameters, cacheable);
+    }
+}

--- a/core/deployment/src/test/java/io/quarkus/deployment/recording/JobParameter.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/recording/JobParameter.java
@@ -1,0 +1,85 @@
+package io.quarkus.deployment.recording;
+
+import java.util.Objects;
+
+public class JobParameter {
+    public static final JobParameter JobContext = new JobParameter(JobParameter.class);
+
+    private String className;
+    private String actualClassName;
+    private Object object;
+
+    private JobParameter() {
+        // used for deserialization
+    }
+
+    private JobParameter(Class<?> clazz) {
+        this(clazz.getName(), null);
+    }
+
+    public JobParameter(Class<?> clazz, Object object) {
+        this(clazz.getName(), object);
+    }
+
+    public JobParameter(Object object) {
+        this(object.getClass().getName(), object);
+    }
+
+    public JobParameter(String className, Object object) {
+        this(className, isNotNullNorAnEnum(object) ? object.getClass().getName() : className, object);
+    }
+
+    public JobParameter(String className, String actualClassName, Object object) {
+        this.className = className;
+        this.actualClassName = actualClassName;
+        this.object = object;
+    }
+
+    /**
+     * Represents the class name expected by the job method (e.g. an object or an interface)
+     *
+     * @return the class name expected by the job method (e.g. an object or an interface)
+     */
+    public String getClassName() {
+        return className;
+    }
+
+    /**
+     * Represents the actual class name of the job parameter (e.g. an object), this will never be an interface
+     *
+     * @return the actual class name of the job parameter (e.g. an object), this will never be an interface
+     */
+    public String getActualClassName() {
+        return actualClassName;
+    }
+
+    /**
+     * The actual job parameter
+     *
+     * @return the actual job parameter
+     */
+    public Object getObject() {
+        return object;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (!(o instanceof JobParameter))
+            return false;
+        JobParameter that = (JobParameter) o;
+        return Objects.equals(className, that.className)
+                && Objects.equals(actualClassName, that.actualClassName)
+                && Objects.equals(object, that.object);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(className, actualClassName, object);
+    }
+
+    protected static boolean isNotNullNorAnEnum(Object object) {
+        return object != null && !(object instanceof Enum);
+    }
+}

--- a/core/deployment/src/test/java/io/quarkus/deployment/recording/TestRecorder.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/recording/TestRecorder.java
@@ -59,6 +59,10 @@ public class TestRecorder {
         RESULT.add(bean);
     }
 
+    public void bean(JobDetails jobDetails) {
+        RESULT.add(jobDetails);
+    }
+
     public void bean(TestJavaBeanWithBoolean bean) {
         RESULT.add(bean);
     }

